File: nodejs/src/strategy/CashPayment.ts
import { PaymentStrategy, PaymentResult } from "./PaymentStrategy";

class CashPayment implements PaymentStrategy {
    private cashGiven: number;

    constructor(cashGiven: number) {
        this.cashGiven = cashGiven;
    }

    getName(): string { return "Tiền mặt"; }

    pay(amount: number): PaymentResult {
        if (this.cashGiven < amount) {
            return { success: false, message: `Thiếu ${(amount - this.cashGiven).toLocaleString()} VND` };
        }
        return {
            success: true,
            message: `Thanh toán ${amount.toLocaleString()} VND. Tiền thừa: ${(this.cashGiven - amount).toLocaleString()} VND`,
        };
    }
}

export default CashPayment;

File: spring/demo/src/main/java/com/example/demo/strategy/CashPayment.java
package com.example.demo.strategy;

public class CashPayment implements PaymentStrategy {
    private final int cashGiven;

    public CashPayment(int cashGiven) {
        this.cashGiven = cashGiven;
    }

    @Override
    public String getName() {
        return "Tiền mặt";
    }

    @Override
    public PaymentResult pay(int amount) {
        if (cashGiven < amount) {
            return new PaymentResult(false, "Thiếu " + (amount - cashGiven) + " VND");
        }

        return new PaymentResult(true, String.format("Thanh toán %,d VND. Tiền thừa: %,d VND", amount, cashGiven - amount));
    }
}
